docs/structured-streaming-programming-guide.md

```diff
@@ -2145,6 +2145,10 @@ streamingDatasetOfString.writeStream.foreach(
     def close(errorOrNull: Throwable): Unit = {
       // Close the connection
     }
+
+    def abort(): Unit = {
+      // Close the connection, this method is optional
```

Comment:
well, it's optional since for backward compatibility i added default empty implementation, but maybe it will be netter to remove the "optional" from the comment

Reply:
Since the original exception is not reaching abort I started to agree with @arunmahadevan.
With a generic exception close can be called.

Reply:
that was my original thought but i didn't want to call close twice each time with a different exception 
will it make sense to add a flag to the ForeachDataWriter which will indicate if the writer has been closed (close has been called) and invoke the close on abort only if closed hasn't need called?

Reply:
Having 2 different functions where in one or both of them releasing allocated resources should happen would make applications horror complicated as a user. Having default implementation for the second function is even worse because resources has to be released maybe in a function which is not known by the app developer (in case of Spark upgrade most of the time devs don't analyze through their code if it compiles and runs somehow).

Reply:
I'm ok with calling close() instead of abort() but i don't think we should call close twice on each error (each with different exception), so my suggestion is to call close with generic exception only if close wasn't called yet

Reply:
Since in the abort case there is no valid exception it's fine to use a generic one.
If I understand it correctly you would like to use this generic exception only. When the problem occurred during processing there is a valid and not generic exception which represents value, consuming it and throwing it away is bad direction (if my assumption is correct). Until now `close` may be called several times because the API doesn't guarantee exactly-once.

Reply:
take a look at the current ForeachDataWriter implementation, when processing fails, the exception is consumed by the close method and then thrown again in order to notify the stream it has failed. due to that the abort is being called so if i'll call the close again in the abort, it will be called again, now with a generic exception

Reply:
> i don't think we should call close twice on each error (each with different exception), so my suggestion is to call close with generic exception only if close wasn't called yet

Since `WriteToDataSourceV2Exec` has only one `catchBlock` and no retry we're aiming to call `abort` once, isn't it? Don't really know what kind of further logic you would like to add.
Are you aiming to define `close will be called exactly-once` in the doc?

Reply:
@gaborgsomogyi 
reverted the changes and just added call to close() instead (if hasn't been called yet)